src/taxonomy/import-tags/data/api.test.js

```diff
@@ -0,0 +1,48 @@
+import MockAdapter from 'axios-mock-adapter';
+import { initializeMockApp } from '@edx/frontend-platform';
+import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
+
+import { tagImportMock, taxonomyImportMock } from '../__mocks__';
+
+import {
+  getTaxonomyImportNewApiUrl,
+  getTagsImportApiUrl,
+  importNewTaxonomy,
+  importTags,
+} from './api';
+
+let axiosMock;
+
+describe('import taxonomy api calls', () => {
+  beforeEach(() => {
+    initializeMockApp({
+      authenticatedUser: {
+        userId: 3,
+        username: 'abc123',
+        administrator: true,
+        roles: [],
+      },
+    });
+    axiosMock = new MockAdapter(getAuthenticatedHttpClient());
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should call import new taxonomy', async () => {
+    axiosMock.onPost(getTaxonomyImportNewApiUrl()).reply(201, taxonomyImportMock);
+    const result = await importNewTaxonomy('Taxonomy name', 'Taxonomy description');
+
+    expect(axiosMock.history.post[0].url).toEqual(getTaxonomyImportNewApiUrl());
+    expect(result).toEqual(taxonomyImportMock);
+  });
+
+  it('should call import tags', async () => {
+    axiosMock.onPut(getTagsImportApiUrl(1)).reply(200, tagImportMock);
+    const result = await importTags(1);
+
+    expect(axiosMock.history.put[0].url).toEqual(getTagsImportApiUrl(1));
+    expect(result).toEqual(tagImportMock);
+  });
+});
```

Comment:
I follow this pattern throughout the code. I would like your opinion on my rationale before changing this.
With the [ADR0002](https://github.com/openedx/frontend-template-application/blob/master/docs/decisions/0002-feature-based-application-organization.rst) in mind, I tried to isolate the features in the tests to make it "self-contained".

In this particular case, the UI is in the `taxonomy-menu` feature and the actual import code is in the `taxonomy-import`.
So, I wrote all the tests with these abstractions in mind:
- I test if the UI calls a function (or shows a dialog) that will start a flow from another feature and handle its results (like calling the function to show a toast)
- I test that this function/dialog works and calls the appropriate hooks/apis
- If there is some API hook in the feature, I wrote a unit test to check if this hook makes the correct API function call and side effects (i.e., invalidate other queries)
- And finally, I check if the API function calls make the correct HTTP request with the correct parameters

I saw comments in other PRs wanting to make these tests more "combined" or "user-oriented", using the UI and checking the Axios calls. I understand that, but I found it difficult to implement without having to:
- Break the "feature barrier"
- Duplicating these tests if we reuse some api/hook/feature in more than one place
- Making the tests too complex (the same tests check the UI components, the HTTP calls and responses, react-query invalidations, toast show, alert, etc..)
- Making the tests too coupled and hard to update

The route I took, basically ended with almost all module files with their own test files for their functions, without much integration. This simplifies some code changes on our side:
   - If we change the import flow, we just need to check that the `taxonomy-menu` feature is calling the correct function/component, and we can abstract the actual implementation of it
   - If we change the API, I just need to rewrite the API Tests, not the function/component that prepares the data for it
   - If we change the way we show success/error messages we just need to update the code in the Layout that handles it; not all functions that show success/error messages

We have pros/cons from both approaches, so I will happily follow your lead on the best way to do this for this project: combining the tests, a more unitary approach or a compromise between the two.

Reply:
I understand your concerns. I'll highlight my reasons for recommending testing in a more combined way and perhaps @bradenmacdonald can also give his views on the same. 

- Re: breaking the "feature barrier". If you're importing one feature from another, you've already broken it. If you have three components, A, B and C, where B and C are under A, and B and C are importing from each other, then perhaps you can test at the level of A. Or you can look at reorganising so that there is no cross-importing. If both B and C are using the same component then perhaps those components should be moved to a common location from which both can import the component. 
- Duplicating tests: I think this is a good thing. More on this later.
- More complex tests: I don't think the tests will be more complex. I think for the API responses you can define the responses once and then reuse those across multiple tests. 
- Making the tests too coupled: The tests are coupled if your code is coupled, not otherwise. 

I think one big disadvantage of testing thing in isolation is that these are not isolated components. If you are testing 5 different components and they all include some common components, then it might seem like those common components are being tested 5 times, but I think they *should* be tested 5 times!

Here is an example. You create a shared component, let's call it TagList. This is now used by both TagListDropdown and TagListModal. At one point you are working on TagListModal and need to make changes to the core TagList component. You update it and TagListModal and they work great together. The TagList tests pass because they only test TagList. The TagListModal and TagListDropdown tests also pass because they both only test their component and mock the shared TagList component. In actuality if you were to go to a page that uses TagListDropdown, it would break, because the changes to TagList made them incompatible. If you used more integrated tests, that didn't use mocks then the tests would ideally fail in this scenario. 

I think if two "features" of the app are completely isolated then it's definitely OK to have them be isolated in tests. However, the fact that the UI is in taxonomy-menu and the import code is in taxonomy-import is exactly the reason why they should be tested together, because they might both work and pass tests in isolation and be completely broken in practice.

Reply:
Hi @xitij2000! Thank you for taking the time to provide a detailed response!

> If you're importing one feature from another, you've already broken it.
I think it is okay to import from siblings from the ADR. I was concerned about putting the tests of feature A in feature B, but I got the overall idea. Maybe the `taxonomy-menu` here is not the best example (because it's just a menu and unlikely to fail), but your TagList example makes a lot of sense.

I will work to write the test this way for the following features.

Let me know if I should fix it here (and in the other PRs). I'm concerned about the budget, but if you think it is worth it, I'm okay with rewriting the tests to improve them.

Again, thank you for the explanation!

Reply:
I tend to agree with @xitij2000 's overall philosophy here. I personally tend to write tests that use (coupled) components exactly as they are used in the app, because (A) it can find bugs with the actual integration of the components, like @xitij2000 explained with that example, (B) it's often much easier to do so than to figure out how to use mocks to isolate the components, and (C) I've spent a lot of time over the years hunting down and fixing test failures that were caused by the mocks, not the actual components themselves.